git_machete/tests/functional/test_machete.py

```diff
@@ -36,110 +37,128 @@ def __init__(self) -> None:
         self.opt_yes: bool = True
 
 
-class GitAPIState:
-    pulls: List[Dict[str, Any]] = [{'head': {'ref': 'bugfix/remove'}, 'user': {'login': 'github_user'}, 'base': {'ref': 'develop'}, 'number': '15', 'html_url': 'www.github.com'}]
-    user: Dict[str, str] = {'login': 'github_user', 'type': 'User', 'company': 'VirtusLab'}
-    issues: List[Dict[str, Any]] = []
-    remote_branches: List[str] = []
+class MockGithubAPIState:
+    def __init__(self, pulls: List[Dict[str, Any]]) -> None:
+        self.pulls: List[Dict[str, Any]] = pulls
+        self.user: Dict[str, str] = {'login': 'github_user', 'type': 'User', 'company': 'VirtusLab'}
+        self.issues: List[Dict[str, Any]] = []
+
+    def new_request(self) -> "MockGithubAPIRequest":
+        return MockGithubAPIRequest(self)
+
+    def get_issue(self, issue_no: str) -> Optional[Dict[str, Any]]:
+        for index, issue in enumerate(self.issues):
+            if issue['number'] == issue_no:
+                return self.issues.pop(index)
+        return None
+
+    def get_pull(self, pull_no: str) -> Optional[Dict[str, Any]]:
+        for index, pull in enumerate(self.pulls):
+            if pull['number'] == pull_no:
+                return self.pulls.pop(index)
+        return None
+
+
+class ResponseWrapper:
+    def __init__(self, data: Union[List[Dict[str, Any]], Dict[str, Any]]) -> None:
+        self.data: Union[List[Dict[str, Any]], Dict[str, Any]] = data
+
+    def read(self) -> Union[str, bytes]:
+        return json.dumps(self.data).encode()
+
+
+class MockGithubAPIRequest:
 
-    def __init__(self, url: str, headers: Dict[str, str] = None, data: Union[str, bytes, None] = None, method: str = ''):
-        self.url: str = url
+    def __init__(self, git_api_state: MockGithubAPIState) -> None:
+        self.git_api_state: MockGithubAPIState = git_api_state
+
+    def __call__(self, url: str, headers: Dict[str, str] = None, data: Union[str, bytes, None] = None, method: str = '') -> "ResponseWrapper":
+        self.parsed_url: ParseResult = urlparse(url, allow_fragments=True)
+        self.parsed_query: Dict[str, List[str]] = parse_qs(self.parsed_url.query)
         self.status_code: Optional[int] = None
```

Comment:
Hmm I think these attributes should probably be set on the `ResponseWrapper` rather than on `MockGithubAPIRequest`

Reply:
I don't see the use for `status_code` now 🤔

Reply:
hmm, actually i don't see point to have `ResposneWrapper`. Let's drag `read` method into `MockGithubAPIRequest`, make private everything which should not be visible outside and return `self` in `__call__`.

Reply:
Hmm I'd still insist on extracting a separate class for response... rather than mixing request-specific and response-specific functionality in a single class, that's gonna be very confusing (_why is there a `return_code` and `read()` on a Request class_)?